File: app/src/main/java/com/dumuzeyn/mp3player/ExcludedTrackIndex.java
package com.dumuzeyn.mp3player;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

/** In-memory exclusion lookup used by scans of thousands of files. */
final class ExcludedTrackIndex {
    private final Map<String, ExcludedTrack> byIdentity = new HashMap<>();
    private final Map<String, ExcludedTrack> byFingerprint = new HashMap<>();

    ExcludedTrackIndex(List<ExcludedTrack> excluded) {
        for (ExcludedTrack item : excluded) {
            byIdentity.put(item.identityKey, item);
            if (!item.fingerprint.isEmpty()) {
                byFingerprint.put(item.fingerprint, item);
            }
        }
    }

    boolean contains(String identityKey, Track track) {
        if (byIdentity.containsKey(identityKey)
                || byIdentity.containsKey(TrackOrigin.uriIdentity(track.uri))) {
            return true;
        }
        ExcludedTrack sameContent = byFingerprint.get(track.fingerprint);
        return sameContent != null && track.fileSize > 0L
                && sameContent.fileSize == track.fileSize;
    }
}

File: app/src/main/java/com/dumuzeyn/mp3player/ExcludedTrackIndex.kt
package com.dumuzeyn.mp3player

/** In-memory exclusion lookup used by scans of thousands of files. */
class ExcludedTrackIndex(excluded: List<ExcludedTrack>) {
    private val byIdentity = HashMap<String, ExcludedTrack>()
    private val byFingerprint = HashMap<String, ExcludedTrack>()

    init {
        excluded.forEach { item ->
            byIdentity[item.identityKey] = item
            if (item.fingerprint.isNotEmpty()) byFingerprint[item.fingerprint] = item
        }
    }

    fun contains(identityKey: String, track: Track): Boolean {
        if (identityKey in byIdentity || TrackOrigin.uriIdentity(track.uri) in byIdentity) return true
        val sameContent = byFingerprint[track.fingerprint]
        return sameContent != null && track.fileSize > 0L && sameContent.fileSize == track.fileSize
    }

    fun containsIdentity(identityKey: String): Boolean = identityKey in byIdentity
}
